pkg/autodiscovery/listeners/kube_services.go

```diff
@@ -25,16 +25,20 @@ import (
 )
 
 const (
-	kubeServiceAnnotationFormat = "ad.datadoghq.com/service.instances"
+	kubeServiceAnnotationFormat  = "ad.datadoghq.com/service.instances"
+	kubeEndpointAnnotationFormat = "ad.datadoghq.com/endpoints.instances"
 )
 
 // KubeServiceListener listens to kubernetes service creation
 type KubeServiceListener struct {
-	informer   infov1.ServiceInformer
-	services   map[types.UID]Service
-	newService chan<- Service
-	delService chan<- Service
-	m          sync.RWMutex
+	servicesInformer   infov1.ServiceInformer
+	endpointsInformer  infov1.EndpointsInformer
+	services           map[types.UID]Service
```

Comment:
If I can just put in my two cents, https://godoc.org/k8s.io/client-go/tools/cache#Store  with  https://godoc.org/k8s.io/client-go/tools/cache#NewStore can be a good fit to store `Services` and `Endpoints`. It embeds and manages the RWMutex for you.

Reply:
I think this could be very useful, maybe we should include it in a next iteration, WDYT? @clamoriniere @xvello

Reply:
I'm all for DRY, but in this case:
  - other listeners won't be able to use it without tying themselves to the `kubeapiserver` build tag, I'd rather have a no-surprises consistency 
  - for this simple use case, I'm not sure we'd be removing much complexity (trading mutex handling for type assertions, as the store returns `interface{}`s)

Reply:
I need to be more familiar with other listeners internal usages, and how build tag are used.

k8s client-go packages provide a set of useful data-structures, to deal with API server events, cache is one of them, but I was also thinking about: https://godoc.org/k8s.io/client-go/util/workqueue that helps to decouple triggered events and the process: dedoubling events, back-off policy in case of processing error...

Reply:
We use workqueues for other informers, see:
  - https://github.com/DataDog/datadog-agent/blob/master/pkg/util/kubernetes/apiserver/metadata_controller.go
  - https://github.com/DataDog/datadog-agent/blob/master/pkg/util/kubernetes/apiserver/hpa_controller.go

The very low traffic of services only didn't justify the overhead, but endpoints do justify it indeed. @ahmed-mez is that something you could work on with Cedric in a follow-up PR?

Reply:
yes! happy to work on it with @clamoriniere in another PR 👍